url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/core/Link.java

```diff
@@ -0,0 +1,514 @@
+        return link(src, dst, NOWEIGHT);
+    }
+    /**
+     * Creates a weighted link from src to dst with weight widx as int[3].
+     * @param src Source index.
+     * @param dst Destination index.
+     * @param widx Weight index.
+     * @return Weighted link.
+     */
+    public static int[] link(final int src, final int dst, final int widx) {
+        return new int[]{ src, dst, widx };
+    }
+    
+    /**
+     * Allocates a memory block for n links.
+     * @param links Number of links.
+     * @return Allocated memory block.
+     */
+    public static int[] alloc(final int links) {
+        return new int[links * LINK_SIZE];
+    }
+    /**
+     * Returns the source index of a link.
+     * @param buffer Array of links.
+     * @param offset Link offset.
+     * @return Source index.
+     */
+    public static int src(final int[] buffer, final int offset) {
+        return buffer[offset + IDX_SRC];
+    }
+    /**
+     * Returns the destination index of a link.
+     * @param buffer Array of links.
+     * @param offset Link offset.
+     * @return Destination index.
+     */
+    public static int dst(final int[] buffer, final int offset) {
+        return buffer[offset + IDX_DST];
+    }
+    /**
+     * Returns the weight index of a link.
+     * @param buffer Array of links.
+     * @param offset Link offset.
+     * @return Weight index.
+     */
+    public static int weight(final int[] buffer, final int offset) {
+        return buffer[offset + IDX_WEIGHT];
+    }
+    
+    //-------------------------------------------------------------------------
+    
+    /**
+     * Swaps two given links.
+     * @param buffer Array of links.
+     * @param first The first link.
+     * @param second The second link.
+     */
+    public static void swap(final int[] buffer, final int first, final int second) {
+        //
+        final int tmp1 = buffer[first];
+        buffer[first]  = buffer[second];
+        buffer[second] = tmp1;
+        //
+        final int tmp2     = buffer[first + 1];
+        buffer[first + 1]  = buffer[second + 1];
+        buffer[second + 1] = tmp2;
+        //
+        final int tmp3     = buffer[first + 2];
+        buffer[first + 2]  = buffer[second + 2];
+        buffer[second + 2] = tmp3;
+        
+    }
+    /**
+     * Compares two links in dst-major order.
+     * @param buffer Array of links.
+     * @param first The first link.
+     * @param second The second link.
+     * @return SMALLER, BIGGER or EQUAL.
+     */
+    public static int cmpDstMaj(final int[] buffer, final int first, final int second) {
+        //
+        if (buffer[first + IDX_DST] < buffer[second + IDX_DST]) return SMALLER;
+        if (buffer[first + IDX_DST] > buffer[second + IDX_DST]) return BIGGER;
+        if (buffer[first + IDX_SRC] < buffer[second + IDX_SRC]) return SMALLER;
+        if (buffer[first + IDX_SRC] > buffer[second + IDX_SRC]) return BIGGER;
+        //
+        return EQUAL;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 176:181, 192:197.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.